# PYTHON-12. Продвинутые методы работы с данными в Pandas 

## 1. Введение

Цели данного модуля:

* Научиться сортировать данные.
* Получить навыки группировки данных и анализа групп.
* Освоить методы построения сводных таблиц и научиться с их помощью исследовать зависимости в признаках.
* Научиться работать с данными, представленными в виде нескольких таблиц, и познакомиться с методами их объединения.

### ВСПОМНИМ, С КАКИМИ ДАННЫМИ МЫ РАБОТАЕМ

In [251]:
import pandas as pd
import numpy as np

In [252]:
melb_data = pd.read_csv('data/melb_data_fe.csv', sep=',')

In [253]:
melb_df = melb_data.copy()

In [254]:
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

In [255]:
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

Обратите внимание, что наше преобразование столбцов к типам datetime и category «слетело».

Почему это произошло?

Ответ на самом деле очень прост: csv-файл не хранит в себе информацию о типах данных столбцов, поэтому при чтении Pandas автоматически определяет тип данных столбца. Не забывайте об этом, обмениваясь преобразованными данными с вашими коллегами.

### Задание 1.1

Преобразуйте столбец Date в формат datetime и выделите квартал (quarter) продажи объектов недвижимости. Найдите второй по популярности квартал продажи. В качестве ответа запишите число объектов, проданных в этом квартале.

In [256]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)

In [257]:
quarter_df = melb_df['Date'].dt.quarter
quarter_df.value_counts().iloc[1]

4359

### Задание 1.2

Преобразуйте все столбцы, в которых меньше 150 уникальных значений, в тип данных category, исключив из преобразования столбцы Date, Rooms, Bedroom, Bathroom, Car.

В качестве ответа запишите результирующее количество столбцов, которые имеют тип данных category.

In [258]:
melb_df.nunique()

Suburb              120
Rooms                 9
Type                  3
Price              2204
Method                5
SellerG              50
Date                 58
Distance            202
Postcode            198
Bedroom              12
Bathroom              9
Car                  11
Landsize           1448
BuildingArea        602
CouncilArea          33
Lattitude          6503
Longtitude         7063
Regionname            8
Propertycount       311
MeanRoomsSquare    1275
AreaRatio          6521
MonthSale            12
AgeBuilding         152
WeekdaySale           7
StreetType           11
Weekend               2
dtype: int64

In [259]:
def cleaner(data):
    exclude_list = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car']
    for col in melb_df.columns:
        if melb_df[col].nunique() < 150 and col not in exclude_list:
            melb_df[col] = melb_df[col].astype('category')
    return data


melb_df.apply(cleaner)    
display(melb_df)

Suburb  Rooms   Type      Price Method   SellerG       Date  \
0        Abbotsford      2  house  1480000.0      S    Biggin 2016-03-12   
1        Abbotsford      2  house  1035000.0      S    Biggin 2016-04-02   
2        Abbotsford      3  house  1465000.0     SP    Biggin 2017-04-03   
3        Abbotsford      3  house   850000.0     PI    Biggin 2017-04-03   
4        Abbotsford      4  house  1600000.0     VB    Nelson 2016-04-06   
...             ...    ...    ...        ...    ...       ...        ...   
13575         other      4  house  1245000.0      S     Barry 2017-08-26   
13576  Williamstown      3  house  1031000.0     SP  Williams 2017-08-26   
13577  Williamstown      3  house  1170000.0      S     Raine 2017-08-26   
13578  Williamstown      4  house  2500000.0     PI   Sweeney 2017-08-26   
13579    Yarraville      4  house  1285000.0     SP   Village 2017-08-26   

       Distance  Postcode  Bedroom  ...  Longtitude  \
0           2.5      3067        2  ...   144.99840   
1           2.5      3067        2  ...   144.99340   
2           2.5      3067        3  ...   144.99440   
3           2.5      3067        3  ...   144.99690   
4           2.5      3067        3  ...   144.99410   
...         ...       ...      ...  ...         ...   
13575      16.7      3150        4  ...   145.16761   
13576       6.8      3016        3  ...   144.87904   
13577       6.8      3016        3  ...   144.88738   
13578       6.8      3016        4  ...   144.89299   
13579       6.3      3013        4  ...   144.88449   

                       Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
0           Northern Metropolitan           4019        25.200000 -0.231707   
1           Northern Metropolitan           4019        15.800000 -0.327660   
2           Northern Metropolitan           4019        18.750000  0.056338   
3           Northern Metropolitan           4019        15.750000  0.145455   
4           Northern Metropolitan           4019        17.750000  0.083969   
...                           ...            ...              ...       ...   
13575  South-Eastern Metropolitan           7392        12.600000 -0.676093   
13576        Western Metropolitan           6380        16.625000 -0.429185   
13577        Western Metropolitan           6380        15.750000 -0.551601   
13578        Western Metropolitan           6380        17.444444 -0.693060   
13579        Western Metropolitan           6543        12.444444 -0.527426   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
0              3           46           5          St        1  
1              4          116           5          St        1  
2              4          117           0          St        0  
3              4           47           0       other        0  
4              4            2           2          St        0  
...          ...          ...         ...         ...      ...  
13575          8           36           5          Cr        1  
13576          8           22           5          Dr        1  
13577          8           20           5          St        1  
13578          8           97           5          St        1  
13579          8           97           5          St        1  

[13580 rows x 26 columns]

In [260]:
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

In [261]:
n = 0
for type in melb_df.dtypes:
    if type == 'category':
        n += 1

print(n)

10


## 2. Сортировка данных в DataFrame

Часто бывает так, что таблицу или промежуточный результат операций с ней необходимо отсортировать по какому-то критерию. Например, для отчётности вам необходимо предоставить список проданных объектов недвижимости, отсортированный по возрастанию цены или дате продажи. Рассмотрим основные подходы к решению таких задач.

### МЕТОД SORT_VALUES()

Для сортировки значений в DataFrame по значениям одного или нескольких столбцов используется метод sort_values().

Основные параметры метода sort_values():

* by — имя или список имён столбцов, по значениям которых производится сортировка.
* axis — ось, по которой производится сортировка (0 — строки, 1 — столбцы). По умолчанию сортировка производится по строкам.
* ascending — сортировка по возрастанию (от меньшего к большему). По умолчанию параметр выставлен на True, для сортировки по убыванию (от большего к меньшему) необходимо выставить его на False.
* ignore_index — создаются ли новые индексы в таблице. По умолчанию выставлен на False и сохраняет индексы изначальной таблицы.
* inplace — производится ли замена исходной таблицы на отсортированную. По умолчанию параметр выставлен на False, то есть замены не производится. Чтобы переопределить исходную таблицу на отсортированную, необходимо выставить этот параметр на True.

СОРТИРОВКА ПО ЗНАЧЕНИЯМ ОДНОГО СТОЛБЦА

In [262]:
melb_df.sort_values(by='Price', axis=0, ascending=True, ignore_index=False, inplace=False).head(10)

Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7293           Albion      1   unit  185000.0      S  hockingstuart   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
8504    Western Metropolitan           5058        42.000000  1.000000   
7293    Western Metropolitan           2185        14.333333 -0.964017   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
8504           4           47           5          St        1  
7293           8           41           0          Rd        0  
7305           7           46           5          St        1  

[10 rows x 26 columns]

СОРТИРОВКА ПО ЗНАЧЕНИЯМ НЕСКОЛЬКИХ СТОЛБЦОВ

Для сортировки по значениям нескольких столбцов необходимо передать названия этих столбцов в параметр by в виде списка. При этом важно обращать внимание на порядок следования столбцов.

Так, например, отсортируем таблицу сначала по возрастанию расстояния от центра города (Distance), а затем — по возрастанию цены объекта (Price). Для того чтобы вывод был более наглядным, выделим каждую десятую строку из столбцов Distance и Price результирующей таблицы:

In [263]:
melb_df.sort_values(by=['Price', 'Distance']).loc[::10, ['Price', 'Distance']]

Price  Distance
2652     85000.0       6.4
7116    200000.0       7.8
5206    220000.0      11.2
7240    241000.0       8.2
2898    250000.0      13.0
...          ...       ...
13341  4060000.0       6.3
8591   4275000.0       1.5
1094   4700000.0      11.2
5801   5046000.0       6.1
251    5525000.0       6.3

[1358 rows x 2 columns]

КОМБИНИРОВАНИЕ СОРТИРОВКИ С ФИЛЬТРАЦИЕЙ

Предположим, компания McGrath поручила нам восстановить хронологию продаж таунхаусов, у которых площадь участка существенно больше площади здания, чтобы понять, как часто компания справляется с таким сложным видом объектов. Объекты, проданные в один и тот же день, мы бы хотели сортировать по значению коэффициента соотношения площадей. 

Найдём информацию о таунхаусах (Type), проданных компанией (SellerG) McGrath, у которых коэффициент соотношения площадей здания и участка (AreaRatio) меньше -0.8. Результат отсортируем по дате продажи (Date) в порядке возрастания, а после проведём сортировку по убыванию коэффициента соотношения площадей. Также обновим старые индексы на новые, установив параметр ignore_index на True. Для наглядности результата выберем из таблицы только столбцы Data и AreaRatio:

In [264]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

Примечание: старайтесь не сочетать фильтрацию и метод sort_values() с параметром inplace=True, так как в таком случае у вас возникнет предупреждение (warning) SettingWithCopyWarning: melb_df[melb_df['Rooms'] > 5].sort_values(inplace=True, by=['Rooms']) 
C:\Users\Andrey\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingwithCopyWarning:

A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/ indexing. html#retu rning-a-view-versus-a-copy

    return func(*args, **kwargs)

Стоит отметить, что это не ошибка и код в таком случае отработает. Однако, Pandas предупреждает вас о том, что при использовании такого кода дальнейшие результаты работы могут быть неожидаемыми.

Если не вдаваться в подробности, то данное предупреждение предназначено для обозначения операций "цепного присваивания". Это ситуация, в которой вы пытаетесь напрямую изменить подмножество исходных данных. В нашем случае мы пытаемся отсортировать с заменой (об этом говорит параметр inplace=True) с заменой исходной таблицы на отсортированную.

Решается данная проблема очень просто: чтобы не возникало подобных конфликтов, мы используем метод copy(), для явного создания копии отфильтрованного подмножества исходных данных и работаем уже с ней (производим сортировку):

filtered = melb_df[melb_df['Rooms'] > 5].copy()

filtered.sort_values(inplace=True, by=['Rooms'])

filtered.head()

### Задание 2.2

Произведите сортировку столбца AreaRatio по убыванию. При этом индексы полученной таблицы замените на новые. Какое значение площади здания находится в строке 1558? Ответ округлите до целого числа.

In [265]:
int(melb_df.sort_values(
    by='AreaRatio', 
    ignore_index=True, 
    ascending=False
).loc[1558, 'BuildingArea'])

126

### Задание 2.3

Найдите таунхаусы (Type) с количеством жилых комнат (Rooms) больше 2. Отсортируйте полученную таблицу сначала по возрастанию числа комнат, а затем по убыванию средней площади комнат (MeanRoomsSquare). Индексы таблицы замените на новые. Какая цена будет у объекта в строке 18? Ответ запишите в виде целого числа.

In [266]:
townhose_df = melb_df[(melb_df['Type'] == 'townhouse') & (melb_df['Rooms'] > 2)]
townhose_df.head(10)

Suburb  Rooms       Type      Price Method        SellerG  \
21     Abbotsford      3  townhouse   900000.0      S          other   
38   Airport West      3  townhouse   840000.0     PI         Nelson   
42   Airport West      3  townhouse   700000.0      S           Brad   
56   Airport West      3  townhouse   700000.0     PI           Brad   
67   Airport West      3  townhouse   540000.0     SP         Nelson   
68   Airport West      3  townhouse   825000.0     SP          Barry   
86    Albert Park      3  townhouse  1550000.0     VB       Marshall   
153        Altona      3  townhouse   800000.0     SP  hockingstuart   
158        Altona      3  townhouse   705000.0      S  hockingstuart   
162        Altona      3  townhouse   743000.0     SP          other   

          Date  Distance  Postcode  Bedroom  ...  Longtitude  \
21  2017-03-18       2.5      3067        3  ...    144.9959   
38  2016-03-09      13.5      3042        3  ...    144.8780   
42  2017-04-03      13.5      3042        3  ...    144.8837   
56  2016-10-15      13.5      3042        3  ...    144.8870   
67  2016-06-18      13.5      3042        2  ...    144.8802   
68  2016-11-19      13.5      3042        3  ...    144.8776   
86  2016-07-11       3.3      3206        3  ...    144.9598   
153 2016-08-10      13.8      3018        3  ...    144.8143   
158 2017-11-02      13.8      3018        3  ...    144.8181   
162 2016-12-11      13.8      3018        3  ...    144.8160   

                Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
21   Northern Metropolitan           4019        15.750000  1.000000   
38    Western Metropolitan           3464        28.125000 -0.147727   
42    Western Metropolitan           3464        16.750000 -0.281501   
56    Western Metropolitan           3464        16.625000 -0.107383   
67    Western Metropolitan           3464        14.428571 -0.229008   
68    Western Metropolitan           3464        12.750000 -0.483544   
86   Southern Metropolitan           3280        15.750000 -0.701422   
153   Western Metropolitan           5301        15.750000 -0.379310   
158   Western Metropolitan           5301        19.000000 -0.128940   
162   Western Metropolitan           5301        22.000000 -0.090439   

     MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
21           3            7           5          St        1  
38           3            0           2          Rd        0  
42           4            8           0          St        0  
56          10            7           5          St        1  
67           6            6           5          Rd        1  
68          11           56           5          St        1  
86           7           46           0          St        0  
153          8           19           2          St        0  
158         11           47           3          Ct        0  
162         12           46           6          St        1  

[10 rows x 26 columns]

In [267]:
int(townhose_df.sort_values(by=['Rooms', 'MeanRoomsSquare'], ascending=[True, False], ignore_index=True).loc[18, 'Price'])

1300000

## 3. Группировка данных в DataFrame

Одна из основных задач анализа данных — это группировка данных и сравнение показателей в группах. Например, нам необходимо сравнить средний уровень заработной платы в зависимости от пола/уровня образования. Или же мы хотим проследить, какая группа клиентов приносит нам наибольший доход, чтобы направить своё внимание на эту группу.

В некоторых случаях группировки может быть достаточно, чтобы ответить на вопросы бизнеса. В других случаях это может стать первым шагом в более сложном анализе. Так, например, благодаря группировке мы можем выявлять признаки, которые не несут статистической значимости, или признаки, которые вносят наибольший вклад. 

### МЕТОД GROUPBY()

В библиотеке Pandas для группировки данных по одному или нескольким признакам можно использовать метод groupby().

Основные параметры метода groupby():

* by — имя или список имён столбцов, по которым производится группировка.
* axis — ось, по которой производится группировка (0 — строки, 1 — столбцы). По умолчанию группировка производится по строкам.
* as_index — добавляется ли дополнительный индекс к таблице. По умолчанию установлен на True.

Метод groupby() возвращает объект DataFrameGroupBy, который хранит в себе информацию о том, какие строки относятся к определённой группе, и сам по себе не представляет для нас интереса.

Однако к этому объекту можно применять уже знакомые нам агрегирующие методы (mean, median, sum и т. д.), чтобы рассчитывать показатели внутри каждой группы.

ГРУППИРОВКА ДАННЫХ ПО ОДНОМУ КРИТЕРИЮ С ОДНОЙ АГРЕГАЦИЕЙ

Применим агрегирующую функцию среднего к результату работы groupby(). В качестве столбца для группировки возьмём столбец типа объекта недвижимости (Type):

In [268]:
melb_df.groupby(by='Type').mean()

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_16936\1287844458.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  melb_df.groupby(by='Type').mean()


Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

Обратите внимание на структуру получившейся таблицы: теперь на месте индексов стоят значения типа объекта недвижимости Type (house, townhouse, unit).

Как правило, нам не нужна информация обо всех столбцах, поэтому агрегирующие методы можно применять только к интересующему нас столбцу. Например, давайте сравним средние цены на объекты в зависимости от их типа:

In [269]:
melb_df.groupby('Type')['Price'].mean()

Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

Примечание. Обратите внимание, что, так как мы считаем только один показатель (среднее) для одного столбца, в результате мы получаем объект Series.

Из этой маленькой таблицы видно, что наибольшей средней ценой обладают объекты типа house (дома, коттеджи, виллы). Следовательно, можно сделать вывод, что тип постройки является значимым фактором при определении цены объекта недвижимости.

Теперь давайте выясним, какие регионы (Regionname) наиболее удалены от центра Мельбурна.

Для этого найдём минимальное значение расстояния от центра города до объекта в зависимости от его региона. Результат отсортируем по убыванию расстояния:

In [270]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

ГРУППИРОВКА ДАННЫХ ПО ОДНОМУ КРИТЕРИЮ С НЕСКОЛЬКИМИ АГРЕГАЦИЯМИ

Чтобы рассчитать несколько агрегирующих методов, можно воспользоваться методом agg(), который принимает список строк с названиями агрегаций.

Давайте построим таблицу для анализа продаж по месяцам. Для этого найдём количество продаж, а также среднее и максимальное значения цен объектов недвижимости (Price), сгруппированных по номеру месяца продажи (MonthSale). Результат отсортируем по количеству продаж в порядке убывания:

In [271]:
melb_df.groupby('MonthSale')['Price'].agg(
    ['count', 'mean', 'max']
).sort_values(by='count', ascending=False)

count          mean        max
MonthSale                                
8           1850  1.056371e+06  6500000.0
7           1835  9.314698e+05  9000000.0
5           1644  1.097807e+06  8000000.0
6           1469  1.068981e+06  7650000.0
3           1408  1.146762e+06  5600000.0
4           1246  1.050479e+06  5500000.0
9           1188  1.126349e+06  6400000.0
10           854  1.135970e+06  6250000.0
11           750  1.142503e+06  5050000.0
12           725  1.144737e+06  5700000.0
2            333  1.169051e+06  4735000.0
1            278  9.397921e+05  5200000.0

Примечание. Обратите внимание, что, так как мы считаем несколько показателей для одного столбца, в результате мы получаем объект DataFrame.

В результате применения метода agg(), в который мы передали список с названиями интересующих нас агрегирующих функций, мы получаем DataFrame со столбцами count, mean и max, где для каждого месяца рассчитаны соответствующие параметры. Результат сортируем по столбцу count.

Какие интересные выводы можно сделать из этой таблицы?

* Пик продаж приходится на период весна-лето.
* Средняя цена продаваемых объектов относительно стабильна и находится в пределах 1 млн. австралийских долларов с небольшими отклонениями (около 100 тыс. влево и вправо).
* Прослеживается некоторая зависимость между сезоном и максимальной ценой объектов: в месяцы с большим спросом на объекты недвижимости цена также имеет наибольшие показатели. Можно сделать предположение, что это связано с повышением цен на элитные дома в периоды большого спроса.

Примечание. Если вам нужна полная информация обо всех основных статистических характеристиках внутри каждой группы, вы можете воспользоваться методом agg(), передав в качестве его параметра строку 'describe':

In [272]:
melb_df.groupby('MonthSale')['Price'].agg('describe')

count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

После базовых математических функций наиболее частым агрегированием является подсчёт числа уникальных значений. Так, например, мы можем вычислить число уникальных риелторских компаний в зависимости от региона, чтобы понять, в каких регионах конкуренция на рынке недвижимости меньше. Это можно сделать, передав в параметр метода agg() строку 'nunique'. 

Более того, метод agg() поддерживает использование и других функций. Передадим дополнительно встроенную функцию set, чтобы получить множество из агентств недвижимости, которые работают в каждом из регионов:

In [273]:
melb_df.groupby('Regionname')['SellerG'].agg(
    		['nunique', set]
)

nunique  \
Regionname                            
Eastern Metropolitan             26   
Eastern Victoria                 11   
Northern Metropolitan            40   
Northern Victoria                11   
South-Eastern Metropolitan       25   
Southern Metropolitan            38   
Western Metropolitan             34   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Eastern Metropolitan        {Miles, Buckingham, HAR, Jellis, Woodards, Phi...  
Eastern Victoria            {Barry, hockingstuart, HAR, Harcourts, Eview, ...  
Northern Metropolitan       {Raine, LITTLE, Miles, Buckingham, McDonald, S...  
Northern Victoria           {Barry, hockingstuart, HAR, YPA, Raine, McGrat...  
South-Eastern Metropolitan  {O'Brien, HAR, Woodards, Jellis, Thomson, Evie...  
Southern Metropolitan       {Raine, LITTLE, Buckingham, O'Brien, HAR, Will...  
Western Metropolitan        {Raine, McDonald, Sweeney, O'Brien, HAR, Bells...  
Western Victoria                 {hockingstuart, YPA, Raine, other, HAR, Ray}

Как и ожидалось, наименьшая конкуренция в наиболее удалённом регионе Western Victoria, а наибольшая — в центральном районе Northern Metropolitan.

Методы группировки по одному признаку позволяют посмотреть на наши данные «в разрезе» группирующего признака и понять их взаимосвязь. Далее мы рассмотрим, как оценивать взаимосвязь между большим количеством признаков.

Предлагаем вам самим потренироваться в использовании группировки ↓

### Задание 3.1

Сгруппируйте данные по признаку количества комнат и найдите среднюю цену объектов недвижимости в каждой группе. В качестве ответа запишите количество комнат, для которых средняя цена наибольшая.

In [274]:
melb_df.groupby('Rooms')['Price'].mean().sort_values(ascending=False)

Rooms
7     1.920700e+06
5     1.870260e+06
6     1.849366e+06
8     1.602750e+06
4     1.445282e+06
3     1.076081e+06
10    9.000000e+05
2     7.750812e+05
1     4.338245e+05
Name: Price, dtype: float64

### Задание 3.2

Какой регион имеет наименьшее стандартное отклонение по географической широте (Lattitude)?
В качестве ответа запишите название этого региона.

In [275]:
melb_df.groupby('Regionname')['Lattitude'].std().sort_values(ascending=True)

Regionname
Western Victoria              0.011579
Southern Metropolitan         0.043080
Eastern Metropolitan          0.047890
Northern Metropolitan         0.049639
Western Metropolitan          0.051251
South-Eastern Metropolitan    0.073411
Northern Victoria             0.084455
Eastern Victoria              0.147067
Name: Lattitude, dtype: float64

### Задание 3.3

Какая риелторская компания (SellerG) имеет наименьшую общую выручку за период с 1 мая по 1 сентября (включительно) 2017 года?
Для ответа на этот вопрос рассчитайте сумму продаж (Price) каждой компании в заданный период.
Не забудьте перевести даты в формат datetime.

In [276]:
# melb_df.groupby('SellerG')['Price'].sum().sort_values(ascending=False)
melb_df[(melb_df['Date'] >= '2017-05-01') & (melb_df['Date'] <= '2017-09-01')].groupby('SellerG')['Price'].sum().sort_values(ascending=True)

SellerG
LITTLE             2742000.0
Cayzer             4439000.0
Burnham            4550500.0
Moonee             7328000.0
Thomson            8332000.0
Bells              8656000.0
Alexkarbon        10985000.0
McDonald          14637500.0
Rendina           15422276.0
Nick              16890000.0
Douglas           18341000.0
Buckingham        19033000.0
C21               19515000.0
Eview             19791500.0
Collins           20217000.0
Philip            22051800.0
Chisholm          23225000.0
Williams          23297000.0
Love              23365500.0
Purplebricks      23401000.0
O'Brien           23855508.0
HAR               25568000.0
Village           26473000.0
RW                29261000.0
Raine             30687700.0
Stockdale         35409800.0
Sweeney           36882750.0
Gary              39138400.0
Hodges            43231000.0
YPA               46354350.0
Miles             47582000.0
Kay               48569500.0
RT                50498000.0
Brad              55955000.0
Jas   

## 4. Сводные таблицы

Сводная таблица принимает на вход данные из отдельных столбцов и группирует их. В результате получается новая таблица, которая позволяет увидеть многомерное обобщение данных. Таким образом, благодаря сводным таблицам мы можем оценить зависимость между двумя и более признаками данных.

Мы чаще сталкиваемся со сводными таблицами, чем с обычными, в плоском виде, так как сводные таблицы удобнее для анализа и быстрых выводов, а также позволяют увидеть более общие зависимости между признаками, нежели простая группировка данных.

МЕТОД GROUPBY КАК СПОСОБ ПОСТРОЕНИЯ СВОДНЫХ ТАБЛИЦ

На самом деле мы с вами уже строили простейшие одномерные сводные таблицы с помощью метода groupby — мы рассматривали сводную таблицу в контексте группировки по одному признаку. 

Например, мы уже умеем строить таблицу, которая показывает зависимость медианной цены и площади здания от числа комнат:

In [277]:
melb_df.groupby('Rooms')[['Price', 'BuildingArea']].median()

Price  BuildingArea
Rooms                         
1       385000.0         107.0
2       690000.0         126.0
3       950000.0         126.0
4      1285000.0         142.0
5      1660000.0         176.0
6      1800000.0         126.0
7      1496000.0         216.5
8      1515000.0         126.0
10      900000.0         126.0

Также можно построить таблицу, в которой мы будем учитывать не только число комнат, но и тип здания (Type). Для этого в параметрах метода groupby() укажем список из нескольких интересующих нас столбцов.

In [278]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean()

Rooms  Type     
1      house        8.668655e+05
       townhouse    5.927045e+05
       unit         3.899289e+05
2      house        1.017238e+06
       townhouse    7.101585e+05
       unit         6.104905e+05
3      house        1.109233e+06
       townhouse    9.847087e+05
       unit         8.505963e+05
4      house        1.462283e+06
       townhouse    1.217092e+06
       unit         1.037476e+06
5      house        1.877327e+06
       townhouse    1.035000e+06
       unit                  NaN
6      house        1.869508e+06
       townhouse             NaN
       unit         5.200000e+05
7      house        1.920700e+06
       townhouse             NaN
       unit                  NaN
8      house        1.510286e+06
       townhouse             NaN
       unit         2.250000e+06
10     house        9.000000e+05
       townhouse             NaN
       unit                  NaN
Name: Price, dtype: float64

В результате выполнения такого кода мы получаем Series, которая обладает несколькими уровнями индексов: первый уровень — число комнат, второй уровень — тип здания. Такая организация индексов называется иерархической. Вычисление параметра (средней цены) происходит во всех возможных комбинациях признаков.

Для того, чтобы финальный результат был представлен в виде сводной таблицы (первый группировочный признак по строкам, а второй — по столбцам), а не в виде Series с иерархическими индексами, к результату чаще всего применяют метод unstack(), который позволяет переопределить вложенный индекс в виде столбцов таблицы:

In [279]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()

Type          house     townhouse          unit
Rooms                                          
1      8.668655e+05  5.927045e+05  3.899289e+05
2      1.017238e+06  7.101585e+05  6.104905e+05
3      1.109233e+06  9.847087e+05  8.505963e+05
4      1.462283e+06  1.217092e+06  1.037476e+06
5      1.877327e+06  1.035000e+06           NaN
6      1.869508e+06           NaN  5.200000e+05
7      1.920700e+06           NaN           NaN
8      1.510286e+06           NaN  2.250000e+06
10     9.000000e+05           NaN           NaN

В результате мы получаем сводную таблицу, столбцы в которой представляют типы домов (house, townhouse, unit), строки — число комнат, а на пересечении строк и столбцов находится средняя стоимость объекта с такими показателями.

Какие интересные выводы можно сделать из этой таблицы?

* Пропуски в сводной таблице (NaN) говорят о том, что в наших данных нет соответствующих комбинаций признаков. Например, у нас нет информации о ценах на таунхаусы, где количество комнат больше пяти.
* Наибольшей средней стоимостью (2,25 млн. австралийских долларов) обладают объекты типа unit с восемью жилыми комнатами. Наименьшая средняя стоимость — у однокомнатных домов типа unit (чуть меньше 400 тыс. австралийских долларов).
* Сколько бы комнат ни было в доме, цена на объекты типа unit всегда ниже других (за исключением восьмикомнатных объектов).

МЕТОД PIVOT_TABLE ДЛЯ ПОСТРОЕНИЯ СВОДНЫХ ТАБЛИЦ

На самом деле метод groupby редко используется при двух параметрах, так как для построения сводных таблиц существует специальный и более простой метод — pivot_table().

Основные параметры метода pivot_table():

* values — имя столбца, по которому необходимо получить сводные данные, применяя агрегирующую функцию;
* index — имя столбца, значения которого станут строками сводной таблицы;
* columns — имя столбца, значения которого станут столбцами сводной таблицы;
* aggfunc — имя или список имён агрегирующих функций (по умолчанию — подсчёт среднего, 'mean');
* fill_value — значение, которым необходимо заполнить пропуски (по умолчанию пропуски не заполняются).

Давайте построим ту же самую таблицу, но уже с использованием метода pivot_table. В качестве параметра values укажем столбец Price, в качестве индексов сводной таблицы возьмём Rooms, а в качестве столбцов — Type. Агрегирующую функцию оставим по умолчанию (среднее). Дополнительно заменим пропуски в таблице на значение 0. Финальный результат для наглядности вывода округлим с помощью метода round() до целых.

In [280]:
melb_df.pivot_table(
    values='Price',
    index='Rooms',
    columns='Type',
    fill_value=0
).round()

Type       house  townhouse       unit
Rooms                                 
1       866866.0   592705.0   389929.0
2      1017238.0   710158.0   610491.0
3      1109233.0   984709.0   850596.0
4      1462283.0  1217092.0  1037476.0
5      1877327.0  1035000.0        0.0
6      1869508.0        0.0   520000.0
7      1920700.0        0.0        0.0
8      1510286.0        0.0  2250000.0
10      900000.0        0.0        0.0

Несложно понять, что метод pivot_table() имеет преимущество перед группировкой по нескольким критериям. Оно заключается в наличии специальных аргументов для строк и столбцов сводной таблицы, благодаря чему уменьшается вероятность запутаться при построении более сложных (многомерных) сводных таблиц, о которых мы поговорим далее.

А теперь давайте проанализируем продажи в каждом из регионов в зависимости от того, будний был день или выходной. Для этого построим сводную таблицу, в которой строками будут являться названия регионов (Regionname), а в столбцах будет располагаться наш «признак-мигалка» выходного дня (Weekend), который равен 1, если день был выходным, и 0 — в противном случае. В качестве значений сводной таблицы возьмём количество продаж.

In [281]:
melb_df.pivot_table(
    values='Price',
    index='Regionname',
    columns='Weekend',
    aggfunc='count'
)

Weekend                        0     1
Regionname                            
Eastern Metropolitan         447  1024
Eastern Victoria              13    40
Northern Metropolitan       1258  2632
Northern Victoria             11    30
South-Eastern Metropolitan   123   327
Southern Metropolitan       1534  3161
Western Metropolitan         960  1988
Western Victoria               8    24

Из результирующей таблицы можно сделать два вывода:
* Число продаж резко возрастает в выходные вне зависимости от региона (приблизительно в 2-3 раза). То есть вероятность того, что дом продадут в выходные, гораздо выше вероятности, что его продадут в будний день.
* В отдалённых регионах (Victoria) коэффициент роста числа продаж выше, чем в центральных. Если в центральных регионах Metropolitan продажи по выходным в 2-2.5 раза выше, чем по будням, то в регионах Victoria число продаж в выходные вырастает примерно в 3 раза.

Такой рост можно даже попытаться объяснить логически: в выходные дни у людей появляется свободное время, чтобы доехать до отдалённых пригородов с целью покупки дома.

Разберём ещё один пример: найдём, как зависит средняя и медианная площадь участка (Landsize) от типа объекта (Type) и его региона (Regionname). Чтобы посмотреть несколько статистических параметров, нужно передать в аргумент aggfunc список из агрегирующих функций. Построим такую сводную таблицу, где пропущенные значения заменим на 0:

In [282]:
melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
).round(1)

median                   mean                 
Type                        house townhouse unit   house townhouse   unit
Regionname                                                               
Eastern Metropolitan        674.0     233.5  203   717.4     269.4  330.4
Eastern Victoria            843.0       0.0  230  3109.0       0.0  295.3
Northern Metropolitan       459.5     134.0    0   619.2     317.3  495.0
Northern Victoria           724.0       0.0    0  3355.5       0.0    0.0
South-Eastern Metropolitan  630.5     240.0  199   664.3     212.2  357.9
Southern Metropolitan       586.0     246.0    0   569.6     278.9  466.4
Western Metropolitan        531.0     198.0   62   507.9     244.6  557.6
Western Victoria            599.5       0.0    0   655.5       0.0    0.0

Обратите внимание на добавление дополнительных индексов столбцов median и mean. Здесь медианное и среднее значения рассчитаны отдельно для каждой комбинации признаков.

Здесь в глаза бросаются объекты типа house в регионах Eastern Victoria и Northern Victoria — в них среднее и медиана отличаются более чем в три раза. Вероятно, это связано с тем, что в этих районах очень большой разброс цен: есть несколько объектов с гигантской площадью, а остальные объекты имеют небольшую площадь. Из-за этого среднее значение искажается, в то время как медиана нечувствительна к такому разбросу и не искажает результат.

### МНОГОМЕРНЫЕ СВОДНЫЕ ТАБЛИЦЫ

До этого мы рассматривали, как некоторый статистический показатель может зависеть от двух признаков. Однако, как уже упоминалось, сводные таблицы позволяют наблюдать зависимость и от большего числа признаков. Такие сводные таблицы называются многомерными. 

Для того чтобы исследовать зависимость от большего числа признаков, можно передать список признаков в параметр index или параметр columns.

Давайте построим таблицу, в которой по индексам будут располагаться признаки метода продажи (Method) и типа объекта (Type), по столбцам — наименование региона (Regionname), а на пересечении строк и столбцов будет стоять медианная цена объекта (Price):

In [283]:
melb_df.pivot_table(
    values='Price',
    index=['Method','Type'],
    columns='Regionname',
    aggfunc='median',
    fill_value=0
)

Regionname        Eastern Metropolitan  Eastern Victoria  \
Method Type                                                
PI     house                   1244000            780000   
       townhouse                760000                 0   
       unit                     650000                 0   
S      house                   1127000            675000   
       townhouse                828000                 0   
       unit                     645750            492000   
SA     house                    932500            950000   
       townhouse                807500                 0   
       unit                          0                 0   
SP     house                   1050000            672500   
       townhouse                910000                 0   
       unit                     515000            400000   
VB     house                   1100000            712500   
       townhouse                892500                 0   
       unit                     500000                 0   

Regionname        Northern Metropolitan  Northern Victoria  \
Method Type                                                  
PI     house                     900000             500000   
       townhouse                 632500                  0   
       unit                      410000                  0   
S      house                     920000             555000   
       townhouse                 750000                  0   
       unit                      525500                  0   
SA     house                     817500             540000   
       townhouse                 425000                  0   
       unit                      616000                  0   
SP     house                     900000             521000   
       townhouse                 690000                  0   
       unit                      470000                  0   
VB     house                    1050000             690000   
       townhouse                 640000                  0   
       unit                      450000                  0   

Regionname        South-Eastern Metropolitan  Southern Metropolitan  \
Method Type                                                           
PI     house                          865000                1725000   
       townhouse                     1190000                1055000   
       unit                           525000                 571250   
S      house                          883300                1611000   
       townhouse                      875000                1135000   
       unit                           606000                 655000   
SA     house                          880000                1390000   
       townhouse                           0                1141000   
       unit                                0                 580000   
SP     house                          770000                1521750   
       townhouse                      800000                1162500   
       unit                           601000                 550000   
VB     house                          850000                1800000   
       townhouse                           0                1250000   
       unit                           700000                 500000   

Regionname        Western Metropolitan  Western Victoria  
Method Type                                               
PI     house                    870000            630000  
       townhouse                670000                 0  
       unit                     360000                 0  
S      house                    870000            397500  
       townhouse                729000                 0  
       unit                     489000                 0  
SA     house                    772500                 0  
       townhouse                467500                 0  
       unit                     571000                 0  
SP     house                    865000            360000  
       t

Первым индексом в таблице идёт метод продажи здания, далее для метода указывается тип недвижимости. По столбцам расположены регионы. В ячейках таблицы указана медианная цена для каждой такой комбинации.

Такие таблицы уже сложнее читать, однако с помощью них можно более глубоко исследовать закономерности. Например, можно видеть, что вне зависимости от метода продажи и региона цена на объекты типа house практически всегда выше, чем на объекты другого типа.

### ДОСТУП К ДАННЫМ В СВОДНОЙ ТАБЛИЦЕ

Давайте рассмотрим, что собой представляют столбцы сложной сводной таблицы.

Запишем сводную таблицу, которую мы создавали ранее в переменную pivot:

In [284]:
pivot = melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

Выведем её столбцы с помощью атрибута columns:

In [285]:
pivot.columns

MultiIndex([('median',     'house'),
            ('median', 'townhouse'),
            ('median',      'unit'),
            (  'mean',     'house'),
            (  'mean', 'townhouse'),
            (  'mean',      'unit')],
           names=[None, 'Type'])

В результате мы получаем объект MultiIndex. Этот объект хранит в себе шесть комбинаций пар столбцов (два статистических параметра и три типа здания), то есть есть шесть возможных вариантов обращения к столбцам таблицы.

Мультииндексы раскрываются подобно вложенным словарям — по очереди, как матрёшка. Чтобы получить доступ к определённому столбцу, вы должны сначала обратиться к столбцу, который находится уровнем выше.

Так, из таблицы pivot мы можем получить средние значения площадей участков для типа здания unit, просто последовательно обратившись по имени столбцов:

In [286]:
display(pivot['mean']['unit'])

Regionname
Eastern Metropolitan          330.444444
Eastern Victoria              295.333333
Northern Metropolitan         495.026538
Northern Victoria               0.000000
South-Eastern Metropolitan    357.864865
Southern Metropolitan         466.380245
Western Metropolitan          557.637232
Western Victoria                0.000000
Name: unit, dtype: float64

Аналогично производится и фильтрация данных. Например, если нам нужны регионы, в которых средняя площадь здания для домов типа house меньше их медианной площади, то мы можем найти их следующим образом:

In [287]:
mask = pivot['mean']['house'] < pivot['median']['house']
filtered_pivot = pivot[mask]
display(filtered_pivot)

median                       mean              \
Type                   house townhouse unit       house   townhouse   
Regionname                                                            
Southern Metropolitan  586.0     246.0    0  569.643881  278.858824   
Western Metropolitan   531.0     198.0   62  507.883406  244.560669   

                                   
Type                         unit  
Regionname                         
Southern Metropolitan  466.380245  
Western Metropolitan   557.637232

Чтобы получить индексы отфильтрованной таблицы, можно воспользоваться атрибутом index и обернуть результат в список:

In [288]:
print(list(filtered_pivot.index))

['Southern Metropolitan', 'Western Metropolitan']


Таким образом, сводные таблицы изначально кажутся сложной структурой, но на самом деле это обычные DataFrame со вложенными индексами строк или столбцов. 

Умение читать и анализировать сложные сводные таблицы — это важный навык, который помогает проводить углублённый анализ данных.

Примечание. На самом деле мультииндексные таблицы можно создавать и вручную. Давайте посмотрим на синтаксис данной конструкции:

In [289]:
import numpy as np
mser = pd.Series(
    np.random.rand(8),
	index=[['white','white','white','blue','blue','red','red','red'], 
           ['up','down','right','up','down','up','down','left']])
display(mser)

white  up       0.358218
       down     0.618926
       right    0.679415
blue   up       0.688744
       down     0.458970
red    up       0.934161
       down     0.469934
       left     0.667895
dtype: float64

В данном примере мы создаём объект Series со вложенными индексами. Мы передаём в качестве индексов Series вложенный список, где первый список задаёт внешний уровень вложенности, а второй список — внутренний уровень вложенности. Значения Series — случайные числа от 0 до 1, сгенерированные функцией np.random.rand() (ваши значения могут отличаться).

Если посмотреть на индексы Series, можно увидеть, что они являются мультииндексами:

In [290]:
print(mser.index)

MultiIndex([('white',    'up'),
            ('white',  'down'),
            ('white', 'right'),
            ( 'blue',    'up'),
            ( 'blue',  'down'),
            (  'red',    'up'),
            (  'red',  'down'),
            (  'red',  'left')],
           )


Аналогично создаются DataFrame со вложенными признаками (вложенными столбцами) — для этого вложенный список передаётся в параметр columns при инициализации таблицы:

In [291]:
mframe = pd.DataFrame(
    np.random.randn(16).reshape(4,4),
    index=[['white','white','red','red'], ['up','down','up','down']],
    columns=[['pen','pen','paper','paper'],[1,2,1,2]]
)
display(mframe)

pen               paper          
                   1         2         1         2
white up   -0.688293  0.000771  1.958419  0.539527
      down  0.616507  0.409484 -0.983307 -0.228282
red   up    0.667132 -1.054907 -1.182057  1.645304
      down  1.035264 -0.633906 -1.140013 -2.251989

### Задание 4.2

Составьте сводную таблицу, которая показывает зависимость медианной площади (BuildingArea) здания от типа объекта недвижимости (Type) и количества жилых комнат в доме (Rooms). Для какой комбинации признаков площадь здания наибольшая?
В качестве ответа запишите эту комбинацию (тип здания, число комнат) через запятую, без пробелов.

In [292]:
melb_df.pivot_table(
    values='BuildingArea',
    index='Type',
    columns='Rooms',
    aggfunc='median',
    fill_value=0
)

Rooms         1    2    3      4    5    6      7    8    10
Type                                                        
house      126.0  126  126  141.0  177  126  216.5  126  126
townhouse   88.0  114  126  159.5  152    0    0.0    0    0
unit        69.5  110  126  126.0    0  171    0.0  126    0

### Задание 4.3

Составьте сводную таблицу, которая показывает зависимость медианной цены объекта недвижимости (Price) от риелторского агентства (SellerG) и типа здания (Type).
Во вновь созданной таблице найдите агентство, у которого медианная цена для зданий типа unit максимальна. В качестве ответа запишите название этого агентства.

In [293]:
price_sell_table = melb_df.pivot_table(
    values='Price',
    index='SellerG',
    columns='Type',
    aggfunc='median',
    fill_value=0
)

display(price_sell_table)


Type             house  townhouse    unit
SellerG                                  
Alexkarbon     1036250     935000  560000
Barry           830000     727500  502000
Bells           692500     430500  457500
Biggin         1151750     675000  548000
Brad            875000     700000  465000
Buckingham      830000     820000  553000
Burnham         922500          0  305000
Buxton         1323750    1030000  670000
C21             836500     768750  625500
Cayzer         1505000    1450000  707500
Chisholm       1520000     950000  640000
Collins        1566500     930000  592000
Douglas         684000     500000  494500
Eview           760000     597500  602000
Fletchers      1390000    1238000  653000
Gary           1370000     993500  550000
Greg           1330000     900000  547500
HAR             695000          0  518000
Harcourts       740000     653750  500000
Hodges         1500000    1105000  605500
Jas             935000     699000  455000
Jellis         1505000    1052000  591750
Kay            2220000    1365000  695000
LITTLE          970000     800000  447000
Love            780000     846000  440000
Marshall       1975000    1408500  715000
McDonald       1242500          0  502500
McGrath        1250000     810000  542500
Miles          1090000     721500  555000
Moonee          806000     715000  622000
Nelson         1055000     730000  530000
Nick           2025000     780000  900000
Noel           1400500     990000  693250
O'Brien         737504     430000  618500
Philip         1035000     701000  636000
Purplebricks    810000     927500  582500
RT             1640000    1400000  630000
RW              870000     630500  625500
Raine           745000     608000  447500
Ray             900000     650000  497500
Rendina        1190000     750000  459000
Stockdale       715500     550000  410000
Sweeney         850000     728000  381000
Thomson        1177000     500000  495000
Village         973500     727500  420000
Williams       1161000     735000  577000
Woodards       1261000     865500  600000
YPA             596250     498500  443500
hockingstuart  1055000     935000  566000
other           901000     840500  535000

In [297]:
max_unit_price = price_sell_table['unit'].max()
price_sell_table[price_sell_table['unit'] == max_unit_price].index[0]

'Nick'